### ISSUE TITLE
Connected box component

### ISSUE THREAD

**Issue by schobocop** [dynatrace-oss/barista#872]
# Feature Request

## Summary

Component which shows the root cause of an issue, as well as previous and subsequent events/steps in the analyzed process.

## Feature Description

The new feature shows a chain of events/steps and highlights an error/warning.
Each event/step consists of a title, an optional subtitle and an optional icon.
There should be an option to show more information about the event/step by expanding it.
It should be able to handle the following things:

- Title is mandatory
- If the title is too long, it should end in ellipsis.
- Subtitle is optional, in case it is not set the title should be centered vertically to the icon.
- If the subtitle is too long, it should end in ellipsis.
- Icon is optional, in case it is not set, the text will move to the left.
- Multiple sections can be expanded at the same time.
- The events/steps are connected via a connection graphic (see attachments)
- They can only be aligned vertically.
- It doesn't support themes.

An event/step can be:
- Expanded on default
- Collapsed on default
- Disabled -> Icon, text and expand icon should be disabled
- It should be possible to lock it in either state expanded / collapsed
- In case there is no body, it should not be expandable.
- Colored with error and warning colors.

Solution: Please see the attachments for a couple of mock ups.

## Attachments

**Basic idea:**
![basic_root_cause](https://user-images.githubusercontent.com/12202020/79311761-325d7500-7efe-11ea-87f3-298abe99d2ce.png)

**With warning instead of error coloring:**
![warning_root_cause](https://user-images.githubusercontent.com/12202020/79311751-312c4800-7efe-11ea-8a32-55a25f770462.png)

**With everything expanded:**
![all_expanded_root_cause](https://user-images.githubusercontent.com/12202020/79311755-325d7500-7efe-11ea-8c20-fc9baf703d24.png)

**With everything collapsed:**
![all_closed_root_cause](https://user-images.githubusercontent.com/12202020/79312053-a0a23780-7efe-11ea-9b89-b5be93a58e77.png)

**Without any icons:**
![no_icon_root_cause](https://user-images.githubusercontent.com/12202020/79311763-32f60b80-7efe-11ea-9ba2-c9b3d8035f0b.png)

**With 2 events without body (not expandable), one locked and a disabled event:**
![locked_and_disabled_root_cause](https://user-images.githubusercontent.com/12202020/79312315-0393ce80-7eff-11ea-8988-8545e56129b5.png)

**Comment by tomheller** [comment]
Hi @schobocop! Thank you for bringing this up and thank you for the detailed description. 

For me, this looks like a combination of our dt-card, expandable behaviors and some connecting styles (for the O---O connections between the cards). 
My question is, would this be a component where multiple teams would be able to benefit from? 
Would you be willing to contribute this component to the library as well?

**Comment by schobocop** [comment]
Hi @tomheller,

we currently use a dt-tile with O---O connections. However, dt-tile can't expand/collapse and requires a body, which we don't always have, so we have to think up content.

After talking to UX, they believe there is benefit for multiple teams and therefore the wish to add it to Barista. 
I talked with @ffriedl89 about a month ago if something similar is already in the pipeline, which he declined and if you have the resources to do it, which he also declined :)

So the plan is that a team member of I will contribute this component to the library. As soon as discussions about the feature are complete, we will plan it into one of our upcoming sprints.

Best

**Comment by lukasholzer** [comment]
Hi @schobocop,

thank you very much for clarifying this! If you want to contribute this to the library then please make sure that you create an API proposal first before starting with the implementation.

In general, our API proposals should describe how a consumer would use the component. This means a snippet of the HTML layout and the inputs and outputs that would be needed for this kind of component. Please post your ideas to this issue to create a discussion about how the API should look like.

**Comment by schobocop** [comment]
Hi @lukasholzer ,

I'll write down a proposal and post it as soon as I'm done with it.
Thanks for the example!

**Comment by schobocop** [comment]
Hi @lukasholzer,
here is a first API proposal. Let me know what you think:

# Initialization
\<dt-connected-box-group\> holds a list of \<dt-connected-box\>
\<dt-connected-box\> holds the following sections:
- \<dt-connected-box-title\> - The title of the event, needs to be defined. Is text only.
- \<dt-connected-box-subtitle\> - Optional. The subtitle of the event, is below the title. Is text only.
- \<dt-connected-box-icon\> - Optional. An icon on the left side of the event.
- \<dt-connected-box-body\> - Optional. Contains a wide variety of content.

# Example:
```html
<dt-connected-box-group>
    <dt-connected-box [disabled]="false">
        <dt-connected-box-icon>
            <dt-icon name="environment"></dt-icon>
        </dt-connected-box-icon>
        <dt-connected-box-title>Some title<dt-connected-box-title>
        <dt-connected-box-subtitle>Subtitle</dt-connected-box-subtitle>
    </dt-connected-box>
    <dt-connected-box [disabled]="false" color="error">
        <dt-connected-box-icon>
            <dt-icon name="warning"></dt-icon>
        </dt-connected-box-icon>
        <dt-connected-box-title>Some title<dt-connected-box-title>
        <dt-connected-box-subtitle>Subtitle</dt-connected-box-subtitle>
        <dt-connected-box-body [expanded]="true">
            Some body describing the event.
        </dt-connected-box-body>
    </dt-connected-box>
    <dt-connected-box [disabled]="true">
        <dt-connected-box-icon>
            <dt-icon name="jewel"></dt-icon>
        </dt-connected-box-icon>
        <dt-connected-box-title>Some title<dt-connected-box-title>
        <dt-connected-box-subtitle>Subtitle</dt-connected-box-subtitle>
    </dt-connected-box>
</dt-connected-box-group>
```
# Inputs
| Name | Type | Default | Description |
--- | --- | --- | ---
| Disabled | boolean | false | If the event should be disabled or not. |
| Expanded | boolean | false | If the event should be expanded or not. |
| Color | 'default' \| 'error' \| 'warning' | default | Color of the event |

# Outputs
 None

**Comment by tomheller** [comment]
@schobocop The API proposal looks good so far. Based on the screenshots you provided, I guess there should also be a switch if the box is `expandable` at all. 
Because there are disabled expandable boxes and boxes that do not have an expand-trigger.

**Comment by schobocop** [comment]
@tomheller if the box is expandable depends on if there is a \<dt-connected-box-body\> given.
Not sure if it will ever happen that a box is disabled but has a body. But it could happen, then we would see the disabled expand icon.

**Comment by ffriedl89** [comment]
The API proposal looks really good - thx a lot @schobocop. 
One thought that came to my mind. Is it ok if the body is initialized right away even when collapsed? Or should we think about lazy init on the body? 

In that case you would need to have an API like

<dt-connected-box [disabled]="false" color="error">
        <dt-connected-box-icon>
            <dt-icon name="warning"></dt-icon>
        </dt-connected-box-icon>
        <dt-connected-box-title>Some title<dt-connected-box-title>
        <dt-connected-box-subtitle>Subtitle</dt-connected-box-subtitle>
        <ng-template dtConnectedBoxBody [expanded]="true">
            Some body describing the event.
        </ng-template>
    </dt-connected-box>

**Comment by schobocop** [comment]
I see the potential benefit of lazy loading @ffriedl89. However, in our current use case, we don't need it. The data for the bodies of the different boxes, in a single request and there is no need to split it up.

Someone else, however, may benefit from it.

I'm not sure how the current procedure in your repo is. Persinally, I'd stick to the necessary parts which we need for our use case to keep the work minimal and not implement something which is currently not needed. However, if lazy loading the body would fit better to the behavior of other components of the repo, I won't object using lazy loading.

**Comment by ffriedl89** [comment]
I am fine with having it the way you described without lazy behavior. I just wanted to bring it up for discussion and consideration.

**Comment by ffriedl89** [comment]
I think the proposal looks really solid! If you want you can post it in our slack channel for more comments. But you can go ahead and get started with the implementation. Please reach out for pointers on how to get started quickly!

**Comment by schobocop** [comment]
Cool thanks! I'll make sure to have the implementation task in one of our upcoming sprints.
